libsharedringbuffer/include/shared_ringbuffer/shared_ringbuffer.h

```diff
@@ -0,0 +1,187 @@
+    uint32_t read_idx;
+} ring_buffer_t;
+
+/* A ring handle for enqueing/dequeuing into  */
+typedef struct ring_handle {
+    ring_buffer_t *used_ring;
+    ring_buffer_t *avail_ring;
+    /* Function to be used to signal that work is queued in the used_ring */
+    notify_fn notify;
+} ring_handle_t;
+
+void ring_init(ring_handle_t *ring, ring_buffer_t *avail, ring_buffer_t *used, notify_fn notify, int buffer_init);
+void register_notification(ring_handle_t *ring, notify_fn notify);
+
+/**
+ * Check if the ring buffer is empty
+ *
+ * @param ring ring buffer to check
+ *
+ * @return true indicates the buffer is empty, false otherwise.
+ */
+static inline int ring_empty(ring_buffer_t *ring) 
+{
+    return !((ring->write_idx - ring->read_idx) % CONFIG_LIB_SHARED_RINGBUFFER_DESC_COUNT);
+}
+
+/**
+ * Check if the ring buffer is full
+ *
+ * @param ring ring buffer to check
+ *
+ * @return true indicates the buffer is full, false otherwise.
+ */
+static inline int ring_full(ring_buffer_t *ring)
+{
+    return !((ring->write_idx - ring->read_idx + 1) % CONFIG_LIB_SHARED_RINGBUFFER_DESC_COUNT);
+}
+
+static inline void notify(ring_handle_t *ring) 
+{
+    return ring->notify();
+}
+
+/**
+ * Enqueue an element to a ring buffer
+ *
+ * @param ring Ring buffer to enqueue into
+ * @param buffer address into shared memory where data is stored.
+ * @param len length of data inside the buffer above
+ * @param cookie optional pointer to data required on dequeueing. 
+ *
+ * @return -1 when ring is empty, 0 on success
+ */
+static inline int enqueue(ring_buffer_t *ring, uintptr_t buffer, unsigned int len, void *cookie) 
+{
+    if (ring_full(ring)) {
+        ZF_LOGE("Ring full");
+        return -1; 
+    }
+
+    ring->buffers[ring->write_idx % CONFIG_LIB_SHARED_RINGBUFFER_DESC_COUNT].encoded_addr = buffer;
+    ring->buffers[ring->write_idx % CONFIG_LIB_SHARED_RINGBUFFER_DESC_COUNT].len = len;
+    ring->buffers[ring->write_idx % CONFIG_LIB_SHARED_RINGBUFFER_DESC_COUNT].cookie = cookie;
+    ring->write_idx++;
+
+    THREAD_MEMORY_RELEASE();
+
+    return 0;
+}
+
+/**
+ * Dequeue an element to a ring buffer
+ *
+ * @param ring Ring buffer to Dequeue from
+ * @param buffer pointer to the address of where to store buffer address.
+ * @param len pointer to variable to store length of data dequeueing
+ * @param cookie pointer optional pointer to data required on dequeueing. 
+ *
+ * @return -1 when ring is empty, 0 on success
+ */
+static inline int dequeue(ring_buffer_t *ring, uintptr_t *addr, unsigned int *len, void **cookie) 
+{
+    if (ring_empty(ring)) {
+        ZF_LOGF("Ring is empty");
+        return -1;
+    }
+
+    *addr = ring->buffers[ring->read_idx % CONFIG_LIB_SHARED_RINGBUFFER_DESC_COUNT].encoded_addr;
+    *len = ring->buffers[ring->read_idx % CONFIG_LIB_SHARED_RINGBUFFER_DESC_COUNT].len;
+    *cookie = ring->buffers[ring->read_idx % CONFIG_LIB_SHARED_RINGBUFFER_DESC_COUNT].cookie;
+
+    THREAD_MEMORY_RELEASE();
+    ring->read_idx++;
```

Comment:
I'm not sure if it'd end up making a noticeable performance difference, but there's an option for these two lines to be combined into `__atomic_store_n(&ring->read_idx, ring->read_idx+1, __ATOMIC_RELEASE);` which allows the compiler to generate slightly different instructions on architectures like aarch64 where there is a store-release instruction (`stlr`) which means an additional `dmb ish` isn't required as shown in the small assembly examples below which are generated from aarch64-linux-gnu-gcc:
```
  4005a8:       d5033bbf        dmb     ish
  4005ac:       90000462        adrp    x2, 48c000 <object.0+0x28>
  .. <cut 2 unrelated instructions>
  4005b8:       b9400841        ldr     w1, [x2, #8]
  4005bc:       11000421        add     w1, w1, #0x1
  4005c0:       b9000841        str     w1, [x2, #8]

```

```
  4005b0:       b9400801        ldr     w1, [x0, #8]
  4005b4:       11000421        add     w1, w1, #0x1
  4005b8:       889ffc41        stlr    w1, [x2]
```

Reply:
Do not really have a requirement for `__atomic_store_n()`? There is no race for incrementing the value, there is only one writer. Also, does this guarantee that all other writes are finish before also, or do we still need a barrier to play safe to enforce this?

Reply:
It's still a plain store, but with release semantics. So it's not allowed to be re-ordered with any memory operations that come before it in program-order.